settings UI: Improve UI for email changes.

saubhagya-patel opened the pull request:
<!-- Describe your pull request here.-->

Fixes: #31975.
[CZO thread](https://chat.zulip.org/#narrow/stream/101-design/topic/email.20change.20UI/near/1945765)

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

**Screenshots and screen captures:**

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [ ] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [x] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [x] Responsiveness and internationalization.
- [x] Strings and tooltips.
- [x] End-to-end functionality of buttons, interactions and flows.
- [x] Corner cases, error conditions, and easily imagined bugs.
</details>

prakhar1144 commented:
Thanks for the fix! Could you please update your commit message to match the [commit style guidelines](https://zulip.readthedocs.io/en/latest/contributing/version-control.html#commit-messages)?

saubhagya-patel commented:
> Thanks for the fix! Could you please update your commit message to match the [commit style guidelines](https://zulip.readthedocs.io/en/latest/contributing/version-control.html#commit-messages)? 

@prakhar1144, I believe the required change in the commit message is related to the use of uppercase words in the commit summary, and I have addressed that. If any other changes are needed, your feedback would be highly valuable. Thank you.

prakhar1144 commented:
Thanks @saubhagya-patel!

Adding it to my list to review this PR this week.

saubhagya-patel commented:
@prakhar1144, could you please review this PR when you have a moment?
I don’t want to cause any inconvenience, but I'm not sure if you're busy. 
There are no new changes, only conflict resolution. Thanks :)

timabbott commented on web/src/settings_account.ts:
```diff
@@ -575,9 +576,28 @@ export function set_up(): void {
+                        $status_element,
+                        "inline-block",
+                    );
+                }, 500);
```
Why is this a `setTimeout`?

saubhagya-patel replied:
Earlier, the email change process followed the use of `settings_ui.do_settings_change()`, which displayed the confirmation notice as a banner and included a checkmark after a 500ms delay using the `setTimeout` method.

However, the recent changes no longer use `settings_ui.do_settings_change()` due to the confirmation notice being displayed as a banner (with the checkmark) in the old implementation. The [issue](https://github.com/zulip/zulip/issues/31975) specifies updating the confirmation notice to a plain text line.

Therefore, I have attempted to replicate the behavior of `settings_ui.do_settings_change()` by displaying the confirmation notice after a 500ms delay using `setTimeout`.
Thank you.

timabbott replied:
Great, this is a type of reasoned explanation that you should aim to provide for interesting decisions/details like this when opening PRs. It takes practice, but it's really helpful if one can explain the key decisions you made.

timabbott replied:
Reading this more carefully, this `setTimeout` still doesn't make any sense for something coming off a modal. I've removed it. 

I'm not sure whether the 500ms in `do_settings_change` makes sense, entirely; can you start a thread in #frontend with before/after screencasts of what happens if we remove it?

saubhagya-patel replied:
Sure, I have started a discussion [here](https://chat.zulip.org/#narrow/channel/6-frontend/topic/setTimeout.20in.20do_settings_change/near/2032891).

saubhagya-patel commented:
@prakhar1144, could you please review this PR when you have a moment?
Thank you.

prakhar1144 commented:
I'll review it today. Thanks!

prakhar1144 commented:
Bug

* Open two browser windows
* Log into the first as normal user and in another as owner.
* Open this setting in the first window. Turn on the `Prevent users from changing their email address` setting from the other window.
* Look at the first window, the tooltip "Email address changes are disabled in this organization." is not visible until you reload.

saubhagya-patel commented on web/templates/settings/account_settings.hbs:
```diff
@@ -9,9 +9,14 @@
+                            <input type="email" value="{{current_user.delivery_email}}" class="button rounded" disabled="disabled" />
+                        </div>
+                        {{#if user_can_change_email}}
+                            <button id="change_email_button" type="button" class="button rounded tippy-zulip-delayed-tooltip" data-tippy-content="{{t 'Change your email' }}" {{#unless user_can_change_email}}disabled="disabled"{{/unless}}>
```
Ooh, I apologize for the inconvenience caused. I think this was missed while resolving conflicts. I will keep this in mind in the future. Thank you.

saubhagya-patel commented on web/templates/settings/account_settings.hbs:
```diff
@@ -9,9 +9,14 @@
                 <div class="input-group">
                     <label class="settings-field-label" for="change_email_button">{{t "Email" }}</label>
                     <div id="change_email_button_container" class="{{#unless user_can_change_email}}disabled_setting_tooltip{{/unless}}">
```
I think we need the class here. This is because of the following state:
- When we open two browser windows and log into the first as normal user and in another as owner.
- Turn on the `Prevent users from changing their email address` setting from the owner window.
- In the updated changes, although the tooltip appears for email changes disabled in the normal user's window for once but if the window of the normal user is refreshed, the tooltip disappears.
- The same is the reason for using `hide` class in the button (`#change_email_button`) as `{{#unless user_can_change_email}}hide{{/unless}}`.

I am including a screen recording to demonstrate this state.

<details>
<summary>Screenrecording of above state.</summary>

https://github.com/user-attachments/assets/b457ba01-ab6d-40e7-ba37-36fd9174ef03

</details>

saubhagya-patel commented:
I have updated this PR following the valuable feedback received. The following changes were made:
- Fixed the noop as mentioned in [this](https://github.com/zulip/zulip/pull/32260#discussion_r1862415335) comment.
- Refactored the node tests in the `settings_org.test.cjs` file to align with the changes of adding the "Edit" button.
- Fixed the bug mentioned in [this](https://github.com/zulip/zulip/pull/32260#issuecomment-2506456398) comment.
  <details>
  <summary>Screen recording of resolved bug</summary>
  
   https://github.com/user-attachments/assets/bf819e6c-a2c7-447f-9f8f-c38b34554fcc
   
   </details>

I must say that I didn't initially recognize the bug mentioned in [this](https://github.com/zulip/zulip/pull/32260#issuecomment-2506456398) comment, but it has proven to be a great lesson for my future testing practices.

This PR is now ready for review. Please let me know if any further adjustments are needed. Thank you.

alya commented:
Let's make the email bold in the confirmation notice.

alya commented:
Let's make the email itself be a disabled text field, like the issue indicates. Currently, it's still styled like a button.

saubhagya-patel commented:
Following the review, I have updated this PR with the following changes:

- The updated email is now displayed in bold, as mentioned in [this](https://github.com/zulip/zulip/pull/32260#issuecomment-2518694846) comment. I have included screenshots of the updated confirmation notice.

  <details>
    <summary>Screenshot of updated confirmation notice</summary>

    ![bold confirmation notice](https://github.com/user-attachments/assets/5b5094e6-7729-487f-b4ae-99b28535167b)

    ![Screenshot 2024-12-05 224707](https://github.com/user-attachments/assets/fff09b51-393c-4146-9303-29f06578b23f)

  </details>

- The email field is now styled like a text field, as mentioned in [this](https://github.com/zulip/zulip/pull/32260#issuecomment-2518705839) comment.

  <details>
    <summary>Screenshot of updated email field</summary>

    ![Screenshot 2024-12-05 225112](https://github.com/user-attachments/assets/c3340e11-41f1-4245-bd20-ad9b2410cbac)

  </details>

- I am uncertain about the dimensions of the "Edit" button. Should it be styled to match the dimensions of the email input field? Specifically, would it be appropriate to apply the same height to the "Edit" button?

As mentioned in [this](https://github.com/zulip/zulip/pull/32260#issuecomment-2518705839) comment:
> And then the disabled tooltip should show up on hover (currently requires a click).
  
@alya, could you please confirm if the issue still persists? On my end, the tooltip appears correctly on hover.
Please let me know if any further adjustments are required. Thank you!

alya commented:
I think that's looking good!

The one thing I noted in manual testing is that we should have a regular text cursor over "Email" rather than a hand cursor when the users doesn't have permission to edit the email field. This also needs to be fixed for the name field.

![Screenshot 2024-12-05 at 15 12 13](https://github.com/user-attachments/assets/e251357b-c09b-4ee7-b269-7fa8c29cfa65)

It's not a big deal if those are follow-ups, so tagging for review from @sahil839 .

alya commented:
@prakhar1144 Does the latest set of changes look good? You can tag for integration review if so.

prakhar1144 commented:
Wasn't able to take a look today, tomorrow for sure!

saubhagya-patel commented:
As mentioned in [this](https://github.com/zulip/zulip/pull/32260#issuecomment-2521694570) comment about using a regular text cursor instead of a hand cursor for the "Email" and "Name" fields, I’ve come up with the following approach:

- Introduce a class named `cursor-text` with `cursor: text;`.
- Dynamically apply this class to the labels, similar to how we handle tooltips (e.g. `.disabled_setting_tooltip`).

@alya, I’d like to open a follow-up PR for this. Should I proceed?

Thank you.

prakhar1144 commented:
>  I’d like to open a follow-up PR for this

Yes, you can.

saubhagya-patel commented:
I have opened a PR [#32262](https://github.com/zulip/zulip/pull/32622) to address the issue as stated in [this](https://github.com/zulip/zulip/pull/32260#issuecomment-2521694570) comment. Thank you.

saubhagya-patel commented:
@prakhar1144, please review this PR once you have a moment. Thank you.

prakhar1144 commented on web/templates/settings/account_settings.hbs:
```diff
@@ -8,9 +8,12 @@
-                    <div id="change_email_button_container" class="{{#unless user_can_change_email}}disabled_setting_tooltip{{/unless}}">
-                        <button id="change_email_button" type="button" class="button rounded tippy-zulip-delayed-tooltip" data-tippy-content="{{t 'Change your email' }}" {{#unless user_can_change_email}}disabled="disabled"{{/unless}}>
-                            {{current_user.delivery_email}}
+                    <div>
```
I think we can skip this `<div>`

saubhagya-patel replied:
@prakhar1144, the `<div>` was placed here intentionally because, when removed, the confirmation notice appears to the right of the "Edit" icon. This is because the confirmation notice is rendered with the `inline-block` class, which accounts for the `margin-bottom` of the `.input-group`.

<details>
<summary>Screenshot with <code>&lt;div&gt;</code> removed.</summary>

![image](https://github.com/user-attachments/assets/8b6b170c-a88c-4884-9132-a6042b4fb675)

</details>

However, if we do not apply any class to the confirmation notice, it is rendered as a block-level element. This causes it to fill up the bottom margin space.

<details>
<summary>Screenshot with no class applied to the confirmation notice.</summary>

![image](https://github.com/user-attachments/assets/997894b1-9903-4878-8bc8-e1b070c40d0c)

</details>

I hope I was able to explain the reasoning behind using the `<div>` here.
Please let me know if any further adjustments are needed. Thank you.

prakhar1144 replied:
Cool, makes sense.

prakhar1144 reviewed (commented):
LGTM apart from that small comment.

timabbott commented:
The live-update was bugged, if you actually completed the flow of changing your email without closing the Zulip window. I fixed with the following:

``` diff
diff --git a/web/src/settings_account.ts b/web/src/settings_account.ts
index c5453d874e..64798dd52c 100644
--- a/web/src/settings_account.ts
+++ b/web/src/settings_account.ts
@@ -43,10 +43,12 @@ let password_quality:
 let user_avatar_widget_created = false;
 
 export function update_email(new_email: string): void {
-    const $email_input = $("#change_email_button");
+    const $email_input = $("#email_field_container");
 
     if ($email_input) {
         $email_input.text(new_email);
+        $("#email-change-status").hide();
+        $("#dev-account-settings-status").hide();
     }
 }
 
```

timabbott commented:
I rebased this, fixed the above, and removed the unnecessary `setTimeout` after triggering the email. With those changes, merged. Thanks for the contribution @saubhagya-patel, but please do be more careful with end-to-end testing every aspect of your PRs when asking for review or after making significant refactoring; I shouldn't be able to find bugs like this.

timabbott commented on web/templates/settings/account_settings.hbs:
```diff
@@ -8,11 +8,15 @@
+                            <input type="email" value="{{current_user.delivery_email}}" class="settings_text_input" disabled="disabled" />
+                        </div>
+                        <button id="change_email_button" type="button" class="button rounded tippy-zulip-delayed-tooltip {{#unless user_can_change_email}}hide{{/unless}}" data-tippy-content="{{t 'Change your email' }}">
+                            <i class="fa fa-pencil" aria-hidden="true"></i>
```
`aria-hidden` is wrong for a button whose value is just an icon. Read about ARIA to learn more, but the basic is that we only want `aria-hidden` for icons that are just decorations; those that are functional need a label.

saubhagya-patel replied:
Sure, I will remember this in the future. Thanks.
